0000-classical-stores.md

```diff
@@ -0,0 +1,493 @@
+The main reason not to expand `Measure` directly in this RFC is the complexity; `Measure` is an _exceptionally_ common operation (perhaps the only one that is almost guaranteed to be in a complete quantum program), and it's not yet clear what the full expansion path here should be.
+Allowing a direct measurement into an `Expr` would involve updating every consumer of `CircuitInstruction.clbits` to know that the values may not always be `Clbit` instances any more, which would be too breaking a change.
+For more discussion of potential ways in which `Measure` and `Store` may begin to look different in the future, see the "Give `Store` first-class support" heading.
+
+With the behaviour described in this RFC, the way to store a measurement result to an arbitrary `Expr` will be
+- measure into a `Clbit`
+- use `QuantumCircuit.store` to copy the measured value into the required place
+
+This is more work for users, but is not unreasonable from the perspective of real hardware, where the locations that qubit measurements are stored into typically need more data transfer and synchronisation to be brought into the places where they can have classical computation done with them.
+The current design does not preclude us changing this in the future.
+
+
+## Questions
+
+### Should we add a `params` field to `CircuitInstruction` now or later?
+
+This RFC currently proposes to keep the `params` of `Store` inherent to the object, much like they are with `RZGate` and other parametrised operations.
+This prevents `Store` from being a singleton, and in this case, means that there's more special-case handling needed.
+
+In this RFC, I went with keeping the parameters inherent to `Store` rather than attempting to thread through a `params` field into `CircuitInstruction`.
+I'm worried that trying to do too much there could interact very oddly with the `.clbits` field, and would lead to an odd split where some things were parametrised in `CircuitInstruction` and others not.
+Until it's clear how _all_ instructions should be parametrised, I elected to keep the new object in the same sort of special case as the other classical operations (the control-flow operations).
+
+
+### What does the API for `QuantumCircuit.compose` expand to?
+
+`QuantumCirucit.compose` is in effect an inlining operation.
+It's not clear to me whether it should be possible for a circuit to involve variables whose names conflict in scope with the program it's being inlined into, or what if any automatic variable-renaming methods we should make available.
+
+If there are no conflicts in the variable names, it seems expected to me that the lists of input variables would catenate, and so on.
+Similarly, two declared `Var` instances that conflict (either by having incompatible initialisers, or one being an input, one not, etc) cannot be allowed to coexist, but does this mean that the composition should be rejected, or should we implement a renaming?
```

Comment:
I think how to handle namespace conflicts should be parametrizable to the `compose` method. It might cover a valid usecase glue circuits referring to the save variable.

Reply:
The only time in this system that you could have two circuits really validly referring to the same variable without a redefinition via the initialiser is in the cases:
- base circuit declares the `Var` as an `input`, inlined declares it as input or closure
- base circuit declares the `Var` as a closure, inlined declares it as closure
- base circuit initialises the `Var`, inlined declares it as closure

Those are all largely ok, but given the nature of how closures work, you have to have a reference to the `Var` already, so I don't think there's much problem in `compose`; we can easily enforce that it only works with the same `Var`.

The issues are if I've got two circuits like
```python
base = QuantumCircuit()
a1 = base.add_var("a", types.Bool(), expr.value(True))

inlined = QuantumCircuit(
a2 = inlined.add_var("a", types.Bool(), expr.value(False))
```
now there are two variables named "a" that are not the same, and naive inlining re-using the same named storage location would cause `a2` to shadow `a1`, and for the ensuing data model to work, we'd already have to rewrite all expressions in `inlined` to use `a1` instead of `a2`.

My preference here would likely be to reject compositions with any variable clashes that would require complete variable rewriting, just for ease in the initial implementation; I think there's nuances to classical-function inlining that I'm not super confident of at the moment to allow something we're not 100% of the utility of.